### DISCUSSION TITLE
please help with mapstruct generator, can't get it to work

### DISCUSSION THREAD

**Discussion by shvaber** [mapstruct/mapstruct#3415] [Q&A] [answered]
good day team!
I'm a newbie with Mapstruct and currently I'm upgrading one project, decided to use mapstruct. My problem is that it hangs on code-generation. Did all things I can, tried much versions and variants - nothing helped. May be your quick look will resolve my issues, may be I just need some small help to have it run properly.
Please find in attachment ZIP-file of the project and DAO jar of the entities. Thank's a lot for any help!
[service.zip](https://github.com/mapstruct/mapstruct/files/12917068/service.zip)
[entities-1.0-SNAPSHOT.jar.gz](https://github.com/mapstruct/mapstruct/files/12917087/entities-1.0-SNAPSHOT.jar.gz)

**Comment by shvaber** [comment]
[build.gradle.zip](https://github.com/mapstruct/mapstruct/files/12917113/build.gradle.zip)

Tried to compile with Gradle - not helped.

**Reply by thunderhook** [reply]
I tried to setup the project and it doesn't work in many ways. The parent pom is missing and I can't set up the entities jar as a dependency. Could you please minify it so that is only a single maven project without a parent pom, so that we can be sure that we're using the same versions?

**Comment by shvaber** [comment]
I'm sorry for parent's dependencies.. Here is a single project, parent is spring-boot-starter-parent
[project.zip](https://github.com/mapstruct/mapstruct/files/12928976/project.zip)

Can you please take a look, is it comfortable to try building?

**Reply by thunderhook** [reply, marked as answer]
I had to change the `<packaging>pom</packaging>` to `jar`, then I was able to reproduce it. I was able to narrow it down to the `CertificateMapper`, but this is problem is present in many mappers of the project:

```java
@Mapper
public interface CertificateMapper {
	Certificate certificateEntityToCertificate(CertificateEntity ue);
	List<Certificate> certificateEntitiesToCertificates(List<CertificateEntity> uel);
}
```

When you comment out the following fields, the mapper  will be generated:
```java
@Data
@NoArgsConstructor
public class Certificate implements Serializable {
	private static final long serialVersionUID = 2647688650163819773L;

	@EqualsAndHashCode.Include
	private Integer id;

//    private SslCert sslcert;
//    private ExtPackage extPackage;
    private User user;
//    private Contact contact;

//    @JsonIgnore
//    private ExtContract extContract;

    private String dnames;
    private String dcvmethod;
    private Calendar dateCreated;
    private Calendar expired;
    private String csr;
    private String privateKey;
    private String crt;
    private Long providerOrder;
    private Boolean dcv;
    private String status;
}
```

The problem seems to be recursive mapping generations which lead to an endless loop (when you wait long enough it results in a `Internal error in the mapping processor: java.lang.OutOfMemoryError: Java heap space`).

Your target entities are extremely intertwoven, just to note a few examples (there are many more):
`Certificate` has
* `SslCert`, which holds a `Set<Certificate>`
* they each hold a `Set<ExtPackage>`
    * each of them: `Set<Certificate>`
    * and a `Set<ExtContract>`
        * each of them again holds a `Set<Certificate>`...

And so on. I think you should rethink the design of your target Dtos.
Or think about the mapping that way: what would you do in that case if it were you to manually map these into the DTOs, would you endlessly map them? Would you stop at a certain level?

My second advice for you would be: disable the implicit mapping (at first, while you figure out what and how (deep) you want to map) and then possible re-enable de implicit mapping again:

```
@Mapper
public interface CertificateMapper {
	@BeanMapping(ignoreByDefault = true)
	@Mapping(target = "id", source = "id")
	// .. and so on
	Certificate certificateEntityToCertificate(CertificateEntity ue);
	List<Certificate> certificateEntitiesToCertificates(List<CertificateEntity> uel);
}
```

My third advice: do not mix up componentModel = "spring" in combination with `FooMapper INSTANCE = Mappers.get(FooMapper.class)`, either remove the component model or the Mappers.get.

Another advice: each mapper now is living as a standalone and does not know about the other mappers. To reuse mappers, use: `@Mapper(users = {OtherMapper.class, AnotherMapper.class})` and so on, so that each mapper does not generate to logic of child elements for themselves.

Another hint: To avoid mapping of cycles, you can use a context like in [this example](https://github.com/mapstruct/mapstruct-examples/tree/main/mapstruct-mapping-with-cycles/src/main/java/org/mapstruct/example/mapper).

Tooling-Tip: Use the mapstruct plugin for your IDE (intellij plugin is more powerful as I know, but there is also an eclipse plugin)

I have to say, this is a very, very challenging project to introduce MapStruct as a newbie. 😅  I would probably comment out most of the targets DTO fields and then try to map each field. When it works comment in the next field, and so on.
Maybe start with the mapper of the most used inner field, like `Article` for example.

I hope I didn't frustrate you and did help you in any way. 😄

**Comment by shvaber** [comment]
thanks a lot for your help!
I removed some Set<?> fields and did some work by your advices - for now it works! It generates mappers quickly.
Thank you so much!
